README.md
GPS-aware datetime module
=========================

This package provides GPS time conversion utilities, including a
gpstime subclass of the built-in datetime class with the addition of
GPS time parsing and conversion methods.

Leap second data, necessary for GPS time conversion, is expected to be
provided by the core libc Time Zone Database tzdata.  If for some
reason the tzdata leapsecond file is not available, a local cache of
the IETF leap second record will be maintained:

  https://www.ietf.org/timezones/data/leap-seconds.list

The package can be executed as a command-line GPS conversion utility.
that uses the gpstime module is also included.  It is a rough
work-alike to LIGO "tconvert" utility.


debian/changelog
gpstime (0.4.6-1) unstable; urgency=medium

  * rev release

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Sat, 17 Apr 2021 15:31:40 -0700

gpstime (0.4.5-1) unstable; urgency=medium

  * New release
  * drop python2 build

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Thu, 15 Apr 2021 16:38:08 -0700

gpstime (0.4.4-1) unstable; urgency=medium

  * New release

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Wed, 04 Dec 2019 15:26:19 -0800

gpstime (0.4.3-1) unstable; urgency=medium

  * New release

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Wed, 27 Nov 2019 08:00:14 -0800

gpstime (0.4.1-1) unstable; urgency=medium

  * New release
  * fix cli library dependency

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Fri, 01 Nov 2019 16:47:34 -0700

gpstime (0.4-1) unstable; urgency=medium

  * New release
  * drop ietf-leap-seconds packages (dropped upstream)

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Wed, 09 Oct 2019 21:52:37 -0700

gpstime (0.3.2-1) unstable; urgency=medium

  * New minor release

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Wed, 22 Aug 2018 18:03:51 -0700

gpstime (0.3.1-1) unstable; urgency=medium

  * New minor release

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Wed, 10 Jan 2018 10:36:19 -0800

gpstime (0.3-1) unstable; urgency=medium

  * New release

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Wed, 14 Jun 2017 11:58:23 -0700

gpstime (0.2-1) unstable; urgency=medium

  * New release

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Sat, 04 Feb 2017 14:32:46 -0800

gpstime (0.1.2-2) unstable; urgency=medium

  * Add build dependencies on python-{dateutil,requests} for tests

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Thu, 13 Oct 2016 11:05:45 -0700

gpstime (0.1.2-1) unstable; urgency=low

  * New release

 -- Jameson Graef Rollins <jameson.rollins@ligo.org>  Wed, 12 Oct 2016 17:25:30 -0700


test.py
#!/usr/bin/env python

import unittest
import os
import time
import datetime
from dateutil.tz import tzutc, tzlocal

os.environ['IETF_LEAPFILE'] = 'leap-seconds.list'
import gpstime

##################################################

class TestGPStime(unittest.TestCase):

    def roundtrip_gps(self, gps):
        self.assertEqual(gpstime.unix2gps(gpstime.gps2unix(gps)), gps)

    def roundtrip_unix(self, unix):
        self.assertEqual(gpstime.gps2unix(gpstime.unix2gps(unix)), unix)

    ##########

    def test_conversion(self):
        self.assertEqual(gpstime.gps2unix(1133585676), 1449550459)
        self.assertEqual(gpstime.unix2gps(1449550459), 1133585676)
        self.assertEqual(1133585676, gpstime.unix2gps(gpstime.gps2unix(1133585676)))

    def test_conversion_past(self):
        self.assertEqual(gpstime.gps2unix(123456789), 439421586)
        self.assertEqual(gpstime.unix2gps(439421586), 123456789)

    def test_conversion_onleap(self):
        self.assertEqual(gpstime.gps2unix(425520006), 741484798)
        self.assertEqual(gpstime.gps2unix(425520007), 741484799)
        self.assertEqual(gpstime.gps2unix(425520008), 741484799)
        self.assertEqual(gpstime.gps2unix(425520009), 741484800)
        self.assertEqual(gpstime.unix2gps(741484798), 425520006)
        self.assertEqual(gpstime.unix2gps(741484799), 425520007)
        self.assertEqual(gpstime.unix2gps(741484800), 425520009)

    def test_roundtrip(self):
        self.roundtrip_gps(123456789)
        self.roundtrip_unix(439421586)

    @unittest.expectedFailure
    def test_roundtrip_onleap(self):
        self.roundtrip_gps(425520008)

    def test_gpstime_new(self):
        self.assertEqual(gpstime.gpstime(2015, 12, 8, 4, 54, 19, 0, tzutc()).gps(),
                         1133585676.0)

    def test_gpstime_fromdatetime(self):
        dt = datetime.datetime(2015, 12, 8, 4, 54, 19, 0, tzutc())
        self.assertEqual(gpstime.gpstime.fromdatetime(dt).gps(),
                         1133585676.0)

    def test_gpstime_parse_now_roundtrip(self):
        gt0 = gpstime.gpstime.parse()
        gt1 = gpstime.gpstime.parse(gt0.gps())
        self.assertEqual(gt0, gt1)

    def test_gpstime_parse_utc(self):
        self.assertEqual(gpstime.gpstime.parse('Dec 08 2015 04:54:19.2 UTC').gps(),
                         1133585676.2)

    def test_gpstime_parse_utc2(self):
        self.assertEqual(gpstime.gpstime.parse('2014-07-03 17:16:14 UTC').gps(),
                         1088442990.0)

    def test_gpstime_parse_local(self):
        self.assertEqual(gpstime.gpstime.parse('Dec 08 2015 04:54:19.2 PST').gps(),
                         1133614476.2)

    def test_gpstime_parse_gps(self):
        self.assertEqual(gpstime.gpstime.parse(1133585676.2).iso(),
                         '2015-12-08T04:54:19.200000Z')

    def test_gpstime_parse_timestamp(self):
        self.assertEqual(gpstime.gpstime.parse('@1474821047').gps(),
                         1158856264)

    def test_gpstime_parse_tconvert(self):
        self.assertEqual(gpstime.gpstime.tconvert('2015-12-08T04:54:19.200000Z'),
                         gpstime.gpstime.tconvert(1133585676.2))

    def test_gpstime_fromgps(self):
        self.assertEqual(gpstime.gpstime.fromgps(1133585676).iso(),
                         '2015-12-08T04:54:19.000000Z')

    def test_gpstime_fromgps_timestamp(self):
        self.assertEqual(gpstime.gpstime.fromgps(1133585676).timestamp(),
                         1449550459)

    def test_gpstime_tconvert_classmethod(self):
        self.assertEqual(gpstime.gpstime.tconvert(1133585676.2).gps(),
                         1133585676.2)

    def test_gpstime_tconvert_iso(self):
        self.assertEqual(gpstime.tconvert('2015-12-08T04:54:19.200000Z'),
                         1133585676.2)

    def test_gpstime_tconvert_gps(self):
        self.assertEqual(gpstime.tconvert(1133585676.2),
                         '2015-12-08 04:54:19.200000 UTC')

    @unittest.expectedFailure
    def test_gpstime_parse_leap(self):
        self.assertEqual(gpstime.gpstime.parse('Jun 30 1993 23:59:60 UTC').gps(),
                         425520008)

    @unittest.expectedFailure
    def test_gpstime_parse_gps_leap(self):
        self.assertEqual(gpstime.gpstime.parse(425520008).iso(),
                         '1993-06-30T23:59:60.000000Z')

##################################################

if __name__ == '__main__':
    unittest.main(verbosity=5, failfast=False, buffer=True)


gpstime/test.py
import unittest
import os
import time
import datetime
from dateutil.tz import tzutc, tzlocal

import gpstime

##################################################

class TestGPStime(unittest.TestCase):

    def roundtrip_gps(self, gps):
        self.assertEqual(gpstime.unix2gps(gpstime.gps2unix(gps)), gps)

    def roundtrip_unix(self, unix):
        self.assertEqual(gpstime.gps2unix(gpstime.unix2gps(unix)), unix)

    ##########

    def test_conversion(self):
        self.assertEqual(gpstime.gps2unix(1133585676), 1449550459)
        self.assertEqual(gpstime.unix2gps(1449550459), 1133585676)
        self.assertEqual(1133585676, gpstime.unix2gps(gpstime.gps2unix(1133585676)))

    def test_conversion_past(self):
        self.assertEqual(gpstime.gps2unix(123456789), 439421586)
        self.assertEqual(gpstime.unix2gps(439421586), 123456789)

    def test_conversion_onleap(self):
        self.assertEqual(gpstime.gps2unix(425520006), 741484798)
        self.assertEqual(gpstime.gps2unix(425520007), 741484799)
        self.assertEqual(gpstime.gps2unix(425520008), 741484799)
        self.assertEqual(gpstime.gps2unix(425520009), 741484800)
        self.assertEqual(gpstime.unix2gps(741484798), 425520006)
        self.assertEqual(gpstime.unix2gps(741484799), 425520007)
        self.assertEqual(gpstime.unix2gps(741484800), 425520009)

    def test_roundtrip(self):
        self.roundtrip_gps(123456789)
        self.roundtrip_unix(439421586)

    @unittest.expectedFailure
    def test_roundtrip_onleap(self):
        self.roundtrip_gps(425520008)

    def test_gpstime_new(self):
        self.assertEqual(gpstime.gpstime(2015, 12, 8, 4, 54, 19, 0, tzutc()).gps(),
                         1133585676.0)

    def test_gpstime_fromdatetime(self):
        dt = datetime.datetime(2015, 12, 8, 4, 54, 19, 0, tzutc())
        self.assertEqual(gpstime.gpstime.fromdatetime(dt).gps(),
                         1133585676.0)

    def test_gpstime_parse_now_roundtrip(self):
        gt0 = gpstime.gpstime.parse()
        gt1 = gpstime.gpstime.parse(gt0.gps())
        self.assertEqual(gt0, gt1)

    def test_gpstime_parse_utc(self):
        self.assertEqual(gpstime.gpstime.parse('Dec 08 2015 04:54:19.2 UTC').gps(),
                         1133585676.2)

    def test_gpstime_parse_utc2(self):
        self.assertEqual(gpstime.gpstime.parse('2014-07-03 17:16:14 UTC').gps(),
                         1088442990.0)

    def test_gpstime_parse_local(self):
        self.assertEqual(gpstime.gpstime.parse('Dec 08 2015 04:54:19.2 PST').gps(),
                         1133614476.2)

    def test_gpstime_parse_gps(self):
        self.assertEqual(gpstime.gpstime.parse(1133585676.2).gps(),
                         1133585676.2)

    def test_gpstime_parse_gps_iso(self):
        self.assertEqual(gpstime.gpstime.parse(1133585676.2).iso(),
                         '2015-12-08T04:54:19.200000Z')

    def test_gpstime_parse_iso(self):
        self.assertEqual(gpstime.gpstime.parse('2015-12-08T04:54:19.200000Z').iso(),
                         '2015-12-08T04:54:19.200000Z')

    def test_gpstime_parse_iso_local(self):
        self.assertEqual(gpstime.gpstime.parse('2015-12-08T04:54:19.200000').iso(),
                         '2015-12-08T12:54:19.200000Z')

    def test_gpstime_parse_iso_alt(self):
        self.assertEqual(gpstime.gpstime.parse('2015-12-08_04:54:19.200000').iso(),
                         '2015-12-08T12:54:19.200000Z')

    def test_gpstime_parse_timestamp(self):
        self.assertEqual(gpstime.gpstime.parse('@1474821047').gps(),
                         1158856264)

    def test_gpstime_parse_tconvert(self):
        self.assertEqual(gpstime.gpstime.tconvert('2015-12-08T04:54:19.200000Z'),
                         gpstime.gpstime.tconvert(1133585676.2))

    def test_gpstime_fromgps(self):
        self.assertEqual(gpstime.gpstime.fromgps(1133585676).iso(),
                         '2015-12-08T04:54:19.000000Z')

    def test_gpstime_fromgps_timestamp(self):
        self.assertEqual(gpstime.gpstime.fromgps(1133585676).timestamp(),
                         1449550459)

    def test_gpstime_tconvert_classmethod(self):
        self.assertEqual(gpstime.gpstime.tconvert(1133585676.2).gps(),
                         1133585676.2)

    def test_gpstime_tconvert_iso(self):
        self.assertEqual(gpstime.tconvert('2015-12-08T04:54:19.200000Z'),
                         1133585676.2)

    def test_gpstime_tconvert_gps(self):
        self.assertEqual(gpstime.tconvert(1133585676.2),
                         '2015-12-08 04:54:19.200000 UTC')

    @unittest.expectedFailure
    def test_gpstime_parse_leap(self):
        self.assertEqual(gpstime.gpstime.parse('Jun 30 1993 23:59:60 UTC').gps(),
                         425520008)

    @unittest.expectedFailure
    def test_gpstime_parse_gps_leap(self):
        self.assertEqual(gpstime.gpstime.parse(425520008).iso(),
                         '1993-06-30T23:59:60.000000Z')

##################################################

if __name__ == '__main__':
    unittest.main(verbosity=5, failfast=False, buffer=True)


release.py
#!/usr/bin/env python3

import os
import sys
import logging
import argparse
import subprocess
import git

logging.basicConfig(
    format='%(message)s',
    level=os.getenv('LOG_LEVEL', logging.INFO),
)

PACKAGE = subprocess.run(
    [sys.executable, 'setup.py', '--name'],
    check=True,
    capture_output=True,
    text=True,
).stdout.strip()
MODULE = __import__(PACKAGE)
CVERS = MODULE.__version__
VFILE = os.path.join(PACKAGE, '__version__.py')


def vers_from_str(s):
    """Return version info from version string

    Version is assumed to be "semantic" [https://semver.org/] of the
    form:

       major.minor.rev+build

    where major, minor, rev will all be ints, and build will be a str.
    Returns tuple of all components.

    """
    # check if there's a build suffix on the string
    build = ''
    vsplit = s.split('+')
    if len(vsplit) > 1:
        build = '+'+vsplit[1]
    try:
        major, minor, rev = vsplit[0].split('.')
    except ValueError:
        raise ValueError("Could not parse version string: {}".format(s))
    return int(major), int(minor), int(rev), build


def write_vfile(vers):
    with open(VFILE, 'w') as f:
        f.write("__version__ = '{}'\n".format(vers))


parser = argparse.ArgumentParser(
    description="""Make new release of python package.

package: {}
version: {}
  vfile: {}

Determines new version based on release type (major.minor.rev),
updates vfile, commits to git repo and tags, creates sdist and wheel
packages, then creates second commit to update to dev version.
""".format(PACKAGE, CVERS, VFILE),
    formatter_class=argparse.RawDescriptionHelpFormatter,
)
cgroup = parser.add_mutually_exclusive_group(required=True)
cgroup.add_argument(
    'release',
    choices=['major', 'minor', 'rev'],
    nargs='?',
    help="release type",
)
cgroup.add_argument(
    '-s', '--set',
    metavar='VERSION',
    type=str,
    help="set version (no git commits)",
)


def main():
    args = parser.parse_args()

    if args.set:
        vers_from_str(args.set)
        write_vfile(args.set)
        exit()

    release = args.release

    major, minor, rev, build = vers_from_str(CVERS)

    if release == 'major':
        major += 1
        minor = 0
        rev = 0
    elif release == 'minor':
        minor += 1
        rev = 0
    elif release == 'rev':
        rev += 1

    version = '{}.{}.{}'.format(major, minor, rev)

    try:
        resp = input("new {} release: {} -> {}\ntype 'yes' to confirm: ".format(release, CVERS, version))
        if resp != 'yes':
            exit("abort.")
    except KeyboardInterrupt:
        exit("\nabort.")

    tags = subprocess.run(['git', 'tag'], capture_output=True, text=True).stdout.split()
    if version in tags:
        exit("git tag already exists for '{}'; aborting.".format(version))

    logging.info("updating to vfile {}...".format(VFILE))
    write_vfile(version)

    msg = "new {} release {}".format(release, version)

    repo = git.Repo()
    logging.info("git commit...")
    repo.git.commit('-m', msg, VFILE)
    logging.info("git tag '{}'...".format(version))
    repo.git.tag('-m', msg, version)

    logging.info("generating sdist and wheel...")
    out = subprocess.run(
        [sys.executable, 'setup.py', 'sdist', 'bdist_wheel'],
        check=True,
        capture_output=True,
        text=True,
    )
    logging.debug(out.stdout.strip())

    logging.info("update to dev version...")
    dvers = version + '+dev0'
    write_vfile(dvers)
    repo.git.commit('-m', "post release dev versioning", VFILE)


if __name__ == '__main__':
    main()


gpstime/__init__.py
"""datetime with GPS time support

This module primarily provides functions for converting to/from UNIX
and GPS times, as well as a `gpstime` class that directly inherits
from the builtin `datetime` class, adding additional methods for GPS
time input and output.

Leap seconds are expected to be provided by the core libc Time Zone
Database tzdata.  If for some reason the tzdata leapsecond file is not
available, a local cache of the IETF leap second record will be
maintained:

  https://www.ietf.org/timezones/data/leap-seconds.list

KNOWN BUGS: This module does not currently handle conversions of time
strings describing the actual leap second themselves, which are
usually represented as the 60th second of the minute during which the
leap second occurs.

"""
from datetime import datetime
import warnings
import argparse
import subprocess

from dateutil.tz import tzutc, tzlocal

from .__version__ import __version__
from .leaps import LEAPDATA


##################################################

ISO_FORMAT = '%Y-%m-%dT%H:%M:%S.%fZ'

# UNIX time for GPS 0 (1980-01-06T00:00:00Z)
GPS0 = 315964800

##################################################


def unix2gps(unix):
    """Convert UNIX timestamp to GPS time.

    """
    unix = float(unix)
    gps = unix - GPS0
    for leap in LEAPDATA:
        if leap < GPS0:
            continue
        if unix < leap:
            break
        gps += 1
    return gps


def gps2unix(gps):
    """Convert GPS time to UNIX timestamp.

    """
    gps = float(gps)
    unix = gps + GPS0
    for leap in LEAPDATA:
        if leap < GPS0:
            continue
        if unix < leap:
            break
        unix -= 1
    return unix


##################################################


class GPSTimeException(Exception):
    pass


def cudate(string='now'):
    """Parse date/time string to UNIX timestamp with GNU coreutils date

    """
    cmd = ['date', '+%s.%N', '-d', string]
    try:
        ts = subprocess.check_output(cmd, stderr=subprocess.PIPE).strip()
    except subprocess.CalledProcessError:
        raise GPSTimeException("could not parse string '{}'".format(string))
    return float(ts)


def dt2ts(dt):
    """Return UNIX timestamp for datetime object.

    """
    try:
        dt = dt.astimezone(tzutc())
        tzero = datetime.fromtimestamp(0, tzutc())
    except ValueError:
        warnings.warn("GPS converstion requires timezone info.  Assuming local time...",
                      RuntimeWarning)
        dt = dt.replace(tzinfo=tzlocal())
        tzero = datetime.fromtimestamp(0, tzlocal())
    delta = dt - tzero
    return delta.total_seconds()


##################################################


class gpstime(datetime):
    """GPS-aware datetime class

    An extension of the datetime class, with the addition of methods
    for converting to/from GPS times:

    >>> from gpstime import gpstime
    >>> gt = gpstime.fromgps(1088442990)
    >>> gt.gps()
    1088442990.0
    >>> gt.strftime('%Y-%m-%d %H:%M:%S %Z')
    '2014-07-03 17:16:14 UTC'
    >>> gpstime.now().gps()
    1133737481.204008

    In addition a natural language parsing `parse` classmethod returns
    a gpstime object for a arbitrary time string:

    >>> gpstime.parse('2014-07-03 17:16:14 UTC').gps()
    1088442990.0
    >>> gpstime.parse('2 days ago').gps()
    1158440653.553765

    """
    def __new__(cls, *args):
        return datetime.__new__(cls, *args)

    @classmethod
    def fromdatetime(cls, datetime):
        """Return gpstime object from datetime object"""
        tzinfo = datetime.tzinfo
        if tzinfo is None:
            tzinfo = tzlocal()
        cls = gpstime(
            datetime.year, datetime.month, datetime.day,
            datetime.hour, datetime.minute, datetime.second, datetime.microsecond,
            tzinfo,
        )
        return cls

    @classmethod
    def fromgps(cls, gps):
        """Return gpstime object corresponding to GPS time."""
        gt = cls.utcfromtimestamp(gps2unix(gps))
        # HACK: in python3, utcfromtimestamp() seems to floor instead
        # of round the microseconds.  this causes round trips to fail.
        # manually fix microseconds here to the rounded value instead.
        ms = int(round((gps - int(gps))*1000000))
        return gt.replace(microsecond=ms, tzinfo=tzutc())

    @classmethod
    def parse(cls, string='now'):
        """Parse an arbitrary time string into a gpstime object.

        If string not specified 'now' is assumed.  Strings that can be
        cast to float are assumed to be GPS times.  Prepend '@' to a
        float to specify a UNIX timestamp.

        This parse uses the natural lanuage parsing abilities of the
        GNU coreutils 'date' utility.  See "DATE STRING" in date(1)
        for information on possible date/time descriptions.

        """
        if string == 'now':
            return cls.now().replace(tzinfo=tzlocal())
        try:
            gps = float(string)
        except ValueError:
            gps = None
        except TypeError:
            raise TypeError("Time specification must be a string, not {!s}".format(type(string)))
        if gps is not None:
            return cls.fromgps(gps)
        try:
            ts = cudate(string)
        except GPSTimeException:
            # try again in case this was an ISO string using
            # underscore instead of T as the separator
            ts = cudate(string.replace('_', 'T'))
        return cls.fromtimestamp(ts).replace(tzinfo=tzlocal())

    tconvert = parse

    def timestamp(self):
        """Return UNIX timestamp (seconds since epoch)."""
        return dt2ts(self)

    def gps(self):
        """Return GPS time as a float."""
        return unix2gps(self.timestamp())

    def iso(self):
        """Return time in standard UTC ISO format."""
        return self.astimezone(tzutc()).strftime(ISO_FORMAT)


def tconvert(string='now', form='%Y-%m-%d %H:%M:%S.%f %Z'):
    """Reimplementation of LIGO "tconvert" binary behavior

    Given a GPS time string, return the date/time string with the
    specified format.  Given a date/time string, return the GPS time.

    This just uses the gpstime.parse() method internally.

    """
    gt = gpstime.parse(string)
    try:
        float(string)
        return gt.strftime(form)
    except ValueError:
        return gt.gps()


def gpsnow():
    """Return current GPS time as a float.

    """
    return gpstime.utcnow().replace(tzinfo=tzutc()).gps()


parse = gpstime.parse


class GPSTimeParseAction(argparse.Action):
    """gpstime argparse argumention parser Action.

    Parses an arbitrary date/time string into a gpstime object.

    """
    def __call__(self, parser, namespace, values, option_string=False):
        # FIXME: support parsing argparse.REMAINDER values list into a
        # single string
        if isinstance(values, list):
            values = ' '.join(values)
        try:
            gps = parse(values)
        except GPSTimeException:
            parser.error("Could not parse date/time string '{}'".format(values))
        setattr(namespace, self.dest, gps)


gpstime.py
#!/usr/bin/env python
"""datetime with GPS time support

This module primarily provides functions for converting to/from UNIX
and GPS times, as well as a `gpstime` class that directly inherits
from the builtin `datetime` class, adding additional methods for GPS
time input and output.

LEAPSECONDS: GPS time does not honor leap seconds.  In order to
translate between GPS and UNIX time all leap seconds since GPS time
zero need to be accounted for.  In this module, leap seconds are
determined by querying the IETF leap second bulletin:

  https://www.ietf.org/timezones/data/leap-seconds.list

Upon import, this module loads and parses a local cache of the
bulletin file, found in one of the following paths:

  ~/.cache/gpstime/leap-seconds.list
  /var/cache/gpstime/leap-seconds.list

If the found file is out of date the module will throw a warning to
the user.  The data can then be updated by running the
'update_leapdata()' command.  This will download the latest version
from the above URL, and store the updated file in the user cache
location.  System administrators may wish to keep an up-to-date
version at the system cache location, which can be updated by running
'update_leapdata("sys")'.  The environment variable IETF_LEAPFILE may
be used to specify a bulletin file directly and bypass any potential
network access.

KNOWN BUGS: This module does not currently handle conversions of time
strings describing the actual leap second themselves, which are
usually represented as the 60th second of the minute during which the
leap second occurs.

"""

from __future__ import print_function
import os
import sys
import time
import requests
from datetime import datetime
import dateutil
import warnings
import subprocess
from dateutil.tz import tzutc, tzlocal

import pkg_resources
try:
    __version__ = pkg_resources.require('gpstime')[0].version
except:
    __version__ = '?.?.?'

##################################################

ISO_FORMAT = '%Y-%m-%dT%H:%M:%S.%fZ'

##################################################

# UNIX time for GPS 0 (1980-01-06T00:00:00Z)
GPS0 = 315964800
LEAPFILE_IETF = 'https://www.ietf.org/timezones/data/leap-seconds.list'
LEAPFILE_USER = os.path.expanduser('~/.cache/gpstime/leap-seconds.list')
LEAPFILE_SYS = '/var/cache/gpstime/leap-seconds.list'

LEAPDATA, LEAPDATA_EXPIRE = None, None

def __ietf_c_to_unix(c):
    """Convert time since century to time since UNIX epoch

    1900-01-01T00:00:00Z -> 1970-01-01T00:00:00Z

    """
    return int(c) - 2208988800

def __ietf_retrieve(path):
    """Download IETF leap second data to path

    """
    dd = os.path.dirname(path)
    if dd != '' and not os.path.exists(dd):
        os.makedirs(dd)
    r = requests.get(LEAPFILE_IETF)
    r.raise_for_status()
    with open(path+'.tmp', 'wb') as f:
        for c in r.iter_content():
            f.write(c)
    os.rename(path+'.tmp', path)

def __ietf_parse_leapfile(leapfile):
    """Parse leap second data from IETF leap-seconds.list file

    https://www.ietf.org/timezones/data/leap-seconds.list

    """
    data = []
    expire = 0
    if not os.path.isfile(leapfile):
        return data, expire
    with open(leapfile) as f:
        for line in f:
            if line[:2] == '#@':
                expire = __ietf_c_to_unix(line.split()[1])
            elif line[0] == '#':
                pass
            else:
                sl = line.split()
                unix = __ietf_c_to_unix(sl[0])
                offset = int(sl[1])
                #print('%s -> %s, %s %s %s' % (sl[0], unix, offset, datetime.fromtimestamp(unix, tzutc()), sl[3:]))
                if unix > GPS0:
                    data.append((unix, offset))
    return data, expire

def __load_leapdata(update=False):
    """Load IETF leap second data from user or system locations

    If the IETF_LEAPFILE environment variable is specified data will
    be loaded from that path.  Otherwise data will be loaded from the
    first of user or system path location that is not expired.

    """
    global LEAPDATA
    global LEAPDATA_EXPIRE
    env_leapfile = os.getenv('IETF_LEAPFILE')
    if env_leapfile:
        LEAPDATA, LEAPDATA_EXPIRE = __ietf_parse_leapfile(env_leapfile)
        if LEAPDATA_EXPIRE >= time.time():
            warnings.warn("leap second data is expired",
                          RuntimeWarning, stacklevel=1)
        return
    for path in [LEAPFILE_USER, LEAPFILE_SYS]:
        LEAPDATA, LEAPDATA_EXPIRE = __ietf_parse_leapfile(path)
        if LEAPDATA_EXPIRE >= time.time():
            return
    # if we're here no non-expired leap data has been found
    if update:
        update_leapdata()
        if LEAPDATA_EXPIRE < time.time():
            warnings.warn("leap second data is expired and could not be updated from IETF",
                          RuntimeWarning, stacklevel=1)
    else:
        warnings.warn("""Leap second data is expired.
Run 'update_leapdata()' to download the latest bulletin from the IETF""",
                      RuntimeWarning, stacklevel=1)

def update_leapdata(loc='user'):
    """Update lead second data and cache file from online IETF bulletin

    By default this updates the user cache location (see
    LEAPFILE_USER), but the system cache (LEAPFILE_SYS) can attempt to
    be updated by specifying loc='sys'.

    """
    global LEAPDATA
    global LEAPDATA_EXPIRE
    if LEAPDATA_EXPIRE >= time.time():
        return
    leapfile = eval('LEAPFILE_'+loc.upper())
    print('updating leap second data from IETF...', file=sys.stderr)
    __ietf_retrieve(leapfile)
    LEAPDATA, LEAPDATA_EXPIRE = __ietf_parse_leapfile(leapfile)

##################################################

def unix2gps(unix):
    """Convert UNIX timestamp to GPS time.

    """
    unix = float(unix)
    gps = unix - GPS0
    for leap, offset in LEAPDATA:
        if unix < leap:
            break
        gps += 1
    return gps

def gps2unix(gps):
    """Convert GPS time to UNIX timestamp.

    """
    gps = float(gps)
    unix = gps + GPS0
    for leap, offset in LEAPDATA:
        if unix < leap:
            break
        unix -= 1
    return unix

##################################################

class GPSTimeException(Exception):
    pass

def cudate(string='now'):
    """Parse date/time string to UNIX timestamp with GNU coreutils date

    """
    cmd = ['date', '+%s.%N', '-d', string]
    try:
        ts = subprocess.check_output(cmd, stderr=subprocess.PIPE).strip()
    except subprocess.CalledProcessError:
        raise GPSTimeException("could not parse string '{}'".format(string))
    return float(ts)

def dt2ts(dt):
    """Return UNIX timestamp for datetime object.

    """
    try:
        dt = dt.astimezone(tzutc())
        tzero = datetime.fromtimestamp(0, tzutc())
    except ValueError:
        warnings.warn("GPS converstion requires timezone info.  Assuming local time...",
                      RuntimeWarning)
        dt = dt.replace(tzinfo=tzlocal())
        tzero = datetime.fromtimestamp(0, tzlocal())
    delta = dt - tzero
    return delta.total_seconds()

##################################################

class gpstime(datetime):
    """GPS-aware datetime class

    An extension of the datetime class, with the addition of methods
    for converting to/from GPS times:

    >>> from gpstime import gpstime
    >>> gt = gpstime.fromgps(1088442990)
    >>> gt.gps()
    1088442990.0
    >>> gt.strftime('%Y-%m-%d %H:%M:%S %Z')
    '2014-07-03 17:16:14 UTC'
    >>> gpstime.now().gps()
    1133737481.204008

    In addition a natural language parsing `parse` classmethod returns
    a gpstime object for a arbitrary time string:

    >>> gpstime.parse('2014-07-03 17:16:14 UTC').gps()
    1088442990.0
    >>> gpstime.parse('2 days ago').gps()
    1158440653.553765

    """
    def __new__(cls, *args):
        return datetime.__new__(cls, *args)

    @classmethod
    def fromdatetime(cls, datetime):
        """Return gpstime object from datetime object"""
        tzinfo = datetime.tzinfo
        if tzinfo is None:
            tzinfo = tzlocal()
        cls = gpstime(datetime.year, datetime.month, datetime.day,
                      datetime.hour, datetime.minute, datetime.second, datetime.microsecond,
                      tzinfo)
        return cls

    @classmethod
    def fromgps(cls, gps):
        """Return gpstime object corresponding to GPS time."""
        gt = cls.utcfromtimestamp(gps2unix(gps))
        return gt.replace(tzinfo=tzutc())

    @classmethod
    def parse(cls, string='now'):
        """Parse an arbitrary time string into a gpstime object.

        If string not specified 'now' is assumed.  Strings that can be
        cast to float are assumed to be GPS time.  Prepend '@' to
        specify a UNIX timestamp.

        This parse uses the natural lanuage parsing abilities of the
        GNU coreutils 'date' utility.  See "DATE STRING" in date(1)
        for information on possible date/time descriptions.

        """
        if not string:
            string = 'now'
        try:
            gps = float(string)
        except ValueError:
            gps = None
        if gps:
            gt = cls.fromgps(gps)
        elif string == 'now':
            gt = cls.now().replace(tzinfo=tzlocal())
        else:
            ts = cudate(string)
            gt = cls.fromtimestamp(ts).replace(tzinfo=tzlocal())
        return gt

    tconvert = parse

    def timestamp(self):
        """Return UNIX timestamp (seconds since epoch)."""
        return dt2ts(self)

    def gps(self):
        """Return GPS time as a float."""
        return float(unix2gps(self.timestamp()))

    def iso(self):
        """Return time in standard UTC ISO format"""
        return self.strftime(ISO_FORMAT)


def tconvert(string='now', form='%Y-%m-%d %H:%M:%S.%f %Z'):
    """Reimplementation of LIGO "tconvert" binary behavior

    Given a GPS time string, return the date/time string with the
    specified format.  Given a date/time string, return the GPS time.

    This just uses the gpstime.parse() method internally.

    """
    gt = gpstime.parse(string)
    try:
        float(string)
        return gt.strftime(form)
    except ValueError:
        return gt.gps()

##################################################
##################################################

if __name__ != '__main__':

    # if this module is being loaded and not executed, load the leap
    # data with updates disabled and warnings enabled
    __load_leapdata(update=False)

else:

    # otherwise, when being executed as a script, load leap data and
    # update as needed
    __load_leapdata(update=True)

    import argparse
    description = """GPS time conversion

Print local, UTC, and GPS time for the specified time string.
"""
    parser = argparse.ArgumentParser(description=description,
                                     formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument('-v', '--version', action='store_true',
                        help="print version number and exit")
    zg = parser.add_mutually_exclusive_group()
    zg.add_argument('-l', '--local', action='store_const', dest='tz', const='local',
                    help="show only local time")
    zg.add_argument('-u', '--utc', action='store_const', dest='tz', const='utc',
                    help="show only UTC time (default)")
    zg.add_argument('-g', '--gps', action='store_const', dest='tz', const='gps',
                    help="show only GPS time")
    fg = parser.add_mutually_exclusive_group()
    fg.set_defaults(format='%Y-%m-%d %H:%M:%S.%f %Z')
    fg.add_argument('-i', '--iso', action='store_const', dest='format', const=ISO_FORMAT,
                    help="output time in ISO format")
    fg.add_argument('-f', '--format',
                    help="output time format (see strftime behavior: https://docs.python.org/2/library/datetime.html#strftime-and-strptime-behavior)")
    parser.add_argument('time', metavar='TIME', nargs=argparse.REMAINDER, default='now',
                        help="time string, in any format (if none specified, output current time)")
    args = parser.parse_args()

    if args.version:
        print(__version__)
        sys.exit()

    def tzname(tz):
        return datetime.now(tz).tzname()

    try:
        gt = gpstime.parse(' '.join(args.time))
    except GPSTimeException as e:
        sys.exit("Error: {}".format(e))

    if not args.tz:
        ltz = tzlocal()
        utz = tzutc()
        print('{}: {}'.format(tzname(ltz), gt.astimezone(ltz).strftime(args.format)))
        print('{}: {}'.format('UTC', gt.astimezone(utz).strftime(args.format)))
        print('{}: {:.6f}'.format('GPS', gt.gps()))
    elif args.tz == 'gps':
        print('{:.6f}'.format(gt.gps()))
    else:
        if args.tz == 'local':
            tz = tzlocal()
        elif args.tz == 'utc':
            tz = tzutc()
        print('{}'.format(gt.astimezone(tz).strftime(args.format)))
